src/Microsoft.TestPlatform.CrossPlatEngine/MultiTestRunFinalization/MultiTestRunFinalizationManager.cs

```diff
@@ -0,0 +1,118 @@
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="MultiTestRunFinalizationManager"/> class.
+        /// </summary>
+        public MultiTestRunFinalizationManager(ITestPlatformEventSource testPlatformEventSource, params IDataCollectorAttachments[] dataCollectorAttachmentsHandlers)
+        {
+            this.testPlatformEventSource = testPlatformEventSource ?? throw new ArgumentNullException(nameof(testPlatformEventSource));
+            this.dataCollectorAttachmentsHandlers = dataCollectorAttachmentsHandlers ?? throw new ArgumentNullException(nameof(dataCollectorAttachmentsHandlers));
+        }
+
+        /// <inheritdoc/>
+        public async Task FinalizeMultiTestRunAsync(ICollection<AttachmentSet> attachments, IMultiTestRunFinalizationEventsHandler eventHandler, CancellationToken cancellationToken)
+        {
+            await InternalFinalizeMultiTestRunAsync(new Collection<AttachmentSet>(attachments.ToList()), eventHandler, cancellationToken).ConfigureAwait(false);
+        }
+
+        /// <inheritdoc/>
+        public Task<Collection<AttachmentSet>> FinalizeMultiTestRunAsync(ICollection<AttachmentSet> attachments, CancellationToken cancellationToken)
+        {
+            return InternalFinalizeMultiTestRunAsync(new Collection<AttachmentSet>(attachments.ToList()), null, cancellationToken);
+        }
+
+        private async Task<Collection<AttachmentSet>> InternalFinalizeMultiTestRunAsync(Collection<AttachmentSet> attachments, IMultiTestRunFinalizationEventsHandler eventHandler, CancellationToken cancellationToken)
+        {
+            try
+            {
+                testPlatformEventSource.MultiTestRunFinalizationStart(attachments?.Count ?? 0);
+
+                cancellationToken.ThrowIfCancellationRequested();                
+
+                var taskCompletionSource = new TaskCompletionSource<object>();
+                using (cancellationToken.Register(() => taskCompletionSource.TrySetCanceled()))
+                {
+                    Task task = Task.Run(() =>
+                    {
+                        HandleAttachments(attachments, cancellationToken);
+                    });
+
+                    var completedTask = await Task.WhenAny(task, taskCompletionSource.Task).ConfigureAwait(false);
+
+                    if (completedTask == task)
+                    {
+                        await task;
+                        eventHandler?.HandleMultiTestRunFinalizationComplete(attachments);
+                        testPlatformEventSource.MultiTestRunFinalizationStop(attachments.Count);
+                        return attachments;
+                    }
+                    else
+                    {
+                        eventHandler?.HandleLogMessage(ObjectModel.Logging.TestMessageLevel.Informational, "Finalization was cancelled.");
+                        eventHandler?.HandleMultiTestRunFinalizationComplete(null);
```

Comment:
>eventHandler?.HandleMultiTestRunFinalizationComplete(null); [](start = 24, length = 59)

Can we send the current set of attachments in this case?

Reply:
fixed